검색자동완성 UI

sungik-choi opened the pull request:
# [미션] 검색자동완성

[검색자동완성 데모 페이지](https://sungikchoi.com/amazon-clone/resources/html/index.html) : 영단어 1000개

## 파일 구조

```
project
│   README.md
│   server.js : nodejs 로컬 서버
│
└─── resources
     └─── scss
     │      reset.scss
     │      style.scss
     │
     └─── css
     │      reset.css
     │      style.css
     │
     └─── data
     │      localData.json : 자동완성 데이터
     │
     └─── html
     │      search.html
     │
     └─── js
            main.js : 메인 실행 스크립트
            model.js : SearchModel 객체
            view.js : SearchView 객체
            controller.js : SearchController 객체
            util.js : 유틸리티 스크립트
```

## 설명

- MVC 패턴을 잘 적용해보고자 노력했습니다. [링크](https://www.sitepoint.com/mvc-design-pattern-javascript/?fbclid=IwAR2tMiItS6SPScWsXOMTEViiFsAv1mVwPZTDmJo6jb_ugfncirwXYz6KN4w)를 참고했습니다.
- Model은 데이터(json) 조작만, View는 DOM 조작만, Controller는 이벤트를 조작하면서, Model에게서 데이터를 받아 View에게 전달해주는 식으로 설계했습니다.

## 아쉬운 점

> 검색UI와 검색자동완성노출UI를 별개의 prototype패턴의 객체로 각각 구현한다.

- MVC 패턴을 먼저 적용해본다고 Model, View, Controller로 나누는데만 급급한 것 같아 이 부분을 신경을 못쓰고 제작했습니다.

> 검색어 입력View, 검색자동완성View를 따로 만든다.

- 이 부분도 마찬가지...
- 처음에 패턴 자체를 신경 안쓰고, 우선 잘 돌아가게 구현한 후에 나중에 리팩토링을 했다면 더 좋았을 거 같습니다. 객체간의 책임을 제대로 고려하지 못해서 구현 자체도 매끄럽게 하지 못했습니다.

> Event 등록은 delegation방법을 적극 활용한다.

- 이벤트 위임 방식이 별로 사용되지 않은 듯 합니다.

---

- 오늘 퀴즈를 하면서 `this`에 대한 이해도가 많이 부족하다는 걸 느꼈습니다. `controller.js`의 `66`번행 `bind(this)`을 사용한 부분도, 왜 `bind`를 하지 않으면 작동하지 않는지는 이해하지만, MVC 패턴에 더해서 `왜 이런식의 코드를 작성했는지?` 에 대해서는 이해를 하지 못했습니다.
- `controller.js`의 `showList()` 메서드에서, 컨트롤러의 책임이 너무 옅다고 느꼈습니다. 컨트롤러는 모델의 데이터를 뷰가 해석하기 좋게 가공해서 넘겨줘야 하는데, 사실상 메서드를 거치기만 하고, 모델에서 뷰로 직접 넘겨주는 형태가 됐습니다. 뷰가 모델에서 반환한 객체들의 키에 직접적으로 접근해서(키의 이름을 아는 상태로) 사용하고 있습니다. 이런 부분은 어떻게 해결해야 각자의 책임이 잘 드러나는 형태가 될 지 궁금합니다.
- 구현에 급급하다보니 어느 순간부터는 메서드의 역할이 불투명해지고, 버그를 해결하려다 보니 이곳저곳에 무분별하게 조건문을 사용한 것 같아 아쉽습니다.
- 이벤트를 어떻게 주고 받고, 언제 DOM이 조작될지 같은 부분들이 어려웠습니다. 그리고 아직도 제대로 이해하지 못했습니다. 이해가 부족하다보니 버그가 발생했을 때, 의심되는 부분들에 무작정 console.log 찍어보거나 특정 코드 붙여넣기 하는 식으로 해결했었습니다.
- 평소에 사용하던 검색창처럼 자연스럽게 동작하려면 생각보다 많은 노력(?)이 들어가야 된다는 걸 느꼈습니다...

### 미구현 기능

- 최근 검색어 View
- 현재 자동완성창이 닫혀진 후에도 키보드로 이전에 남아있던 목록에 접근이 가능합니다(이동 가능)
- 검색어가 바뀌고 리스트가 초기화돼도 리스트의 인덱스(배경색이 변경되는)가 초기화되지 않고 유지됩니다.

sungik-choi committed:
Docs: 파일 초기화

(cherry picked from commit 0e28b41feb33e585fd4d28576a207f9f231e2bb8)

sungik-choi committed:
Docs/Style: 파일 추가 및 기본 스타일 지정

기본적인 html 구조와 scss로 스타일 작업을 완료했다. 아직 검색 자동완성시 나올 리스트의 스타일은 정의하지 않았다.

(cherry picked from commit c06fafb4404832b3d8a0e1126d4670563be9f43d)

sungik-choi committed:
Style: 검색 자동완성 li 스타일 추가

나중에 검색어가 올라가게 되면 보이게 될 자동완성 리스트의 css를 작성했다. 유지보수가 쉽게 변동가능성이 큰 값들은 scss의 변수로 변경하는 과정을 거쳤다.
현재 HTML의 자동완성 데이터는 임시로 넣어두었고, css는 visibility: hidden으로 설정해두었다.

(cherry picked from commit a34514d884ad2360e0e7fd8ea382c4ccb142e6b2)

sungik-choi committed:
Style/Docs: style 세부 변경 및 json, js 파일 추가

세부적인 스타일을 변경하고, a로 시작하는 영단어 50개로 이루어진 json파일을 추가했다.
MVC 패턴을 공부해보고, 참고하면서 작성해보고자 main.js 파일을 작성중에 있는데, 문득 생각든 점이 아직은 조금 이해하기 이르다. 라는 점이다.
아직 어디에 어떤 개념이 정확히 적용되어야할지 모르기 때문에, 처음부터 역할을 나누지는 말아야겠다.
우선 작동부터 되게 다시 스크립트를 작성해봐야겠다.

(cherry picked from commit 4c4b1739061b979e639883dd3a84bc202c8fc368)

sungik-choi committed:
Style/Feat: li 하이라이트를 위한 코드 변경 및 스크립트 작성

나중에 리스트를 키보드 위아래 키로 움직일 때, 배경색이 바뀌어야 하는 부분을 제대로 표현하기 위해서 li 부분의 구조(마진, 패딩)를 그에 맞게 변경했다.
미리 active클래스들을 추가했다. 이 클래스의 추가와 제거를 통해 스타일을 조작하게 될 것이다. 그 외 작은 시각 보정 작업들을 진행했다.

main.js : Controller 에서 인풋 태그에 키보드를 입력하면 View에게 스타일 변경을 지시하고, Model에 데이터를 불러오게 하는 작업을 진행중이다. 300ms 지연 시간 후에 정보를 요청하도록 해야하는데, 내가 원하는 방식은 연속 입력이 완료된 후, 그 연속 입력된 마지막 텍스트만 잡아서 요청하는 부분인데, 제대로 작동하지 않는다. 지금은 비동기 콜백이 모든 키 입력 이벤트를 300ms 지연 뒤에 전부 다 가져온다.

71번 행의 bind(this) 를 참고해서 작성했는데, 디버깅해보니 이 부분이 없으면 제대로 작동하지 않는다. bind의 필요성을 조금은 알것 같다. this가 제대로 연결되지 않으니, 메서드를 전달해주어도 그 내부에서 this를 참조한다면 에러가 나서 프로그램이 정지된다.

(cherry picked from commit 46dc16e0a33cafdc8d96027060981129a3c32de7)

sungik-choi committed:
Fix: clearTimeout() 을 활용해 타이머 초기화

[레퍼런스](https://schier.co/blog/wait-for-user-to-stop-typing-using-javascript) 사이트를 참고해서 clearTimeout으로 하고자 했던 내용을 구현했다.
setTimeout 전체를 timeout이라는 변수 안에 담고, 만약 다른 키가 입력됐을 때 아직 지연시간이 지나지 않아 실행되지 않았다면 (이 말은 즉, 지연시간 내에 연속입력 했다면 과 같다)
clearTimeout()으로 타이머를 비워버리고 다시 새로운 타이머를 등록한다.

익숙해지게 노력해봐야겠다.

(cherry picked from commit 3f79b501ab117d7a8166d5a205ee9fcaa2e47836)

sungik-choi committed:
Feat/Fix: 자동완성 기능 추가중

여러 고민들 끝에 자동완성 기능을 추가하는 중이다. 현재 문제점이 있는데, 해당되는 문자가 없을 때, active 스타일이 .line에 적용되있던게 사라지지 않아서 굉장히 이상하게 보인다.
이 부분을 해결하고자 여러 시도를 해봤는데, input event의 addClass와 이상하게 겹치면서 깜빡이는 현상이 발생한다.
Model로 부터 온 word가 없다면 active 클래스를 추가하지 않으면 되는데, 이 부분이 생각보다 어렵다. fetch되는 부분에서 bind를 쓰지 않고 return값 등으로 받아보려고 했는데,
왜 bind를 썼는지 알 것 같았다. 왜인지 값이 undefined로 전달된다. 확실히 fetch는 then 블록 안에서 일을 끝마쳐야 하는가보다.
단순히 작동만 하게 하면 할 수 있을 거 같은데, 일관성을 유지하면서 작동하게 하려고 하니 쉽지 않다.

(cherry picked from commit 43976f29dccd7285fde52ade04a69231146b5a2e)

sungik-choi committed:
Fix: 정규표현식 관련 문제 해결

기존에 tempalting에서 오는 word를 현재 입력한 단어로 split 하다보니, 이상하게 단어가 잘리는 현상이 발생하고 있었다.
모델에서 전체 단어를 던져주는게 아니라, 정규표현식의 캡쳐링을 이용해서 일치하는 단어 / 나머지 단어(굵게 표시할) 를 나눠서 던져주고,
뷰에서 객체를 해석하는 식으로 해결했다.

조금 의구심이 드는건, 이런 형태라면 Controller의 역할이 이 부분에서는 너무 적다고 느껴진다. Model과 View의 의존성이 너무 큰 느낌?
Model에서 받은 객체를 Controller가 더 해석해서 넘겨주는 방향을 고민해봐야겠다.

(cherry picked from commit 4d8a8b959f6c755701c4e1e9974f35648054adb4)

sungik-choi committed:
Fix: 자동완성 스타일 변경 문제 수정 및 변수명 일부 수정

깔끔하지는 못한 해결방법 같지만, 컨트롤러의 showList 메서드에 뷰의 addClass 메서드를 추가하는 방향으로 해결했다. 템플레이팅 된 후, 단어들의 길이가 0이상이 아니라면 클래스를 더하지 않도록.
114번행의 클래스명 변수가 중복돼서 조금 마음에 안들지만 우선은 작동해서 다행이다.

뷰의 기본값을 ||로 설정해주려고 했는데, 아예 입력을 안하면 undefined가 나오며 멈춰버린다. 생성자 내에서 기본 파라미터로 설정해줘봤는데, 맞는 방법인지 모르겠다.

(cherry picked from commit 8b9a046b1e2bf8ffba812e870dd9991c7dd6c799)

sungik-choi committed:
Docs: json 데이터 추가

정확한 테스트를 위해서 40개에서 1000개로 늘렸다.

(cherry picked from commit 954bb9c3e7de9f22da34bbb6533ff044c0175d74)

sungik-choi committed:
Feat/Style: 자동완성 리스트 하이라이트 추가

리스트가 나오고 화살표로 이동하면 자동완성 리스트가 하이라이트 되게 변경했다. 위아래로 움직이면 해당 내용이 input 태그의 value로 들어가고, 엔터를 누르거나 검색 아이콘을 클릭하면 자동검색창은 닫힌다.
하지만 깔끔하게 작동하지 않는다... 현재 데이터를 패치해오고 리스트가 변경됐을때 인덱스또한 초기화돼야하는데 현재 미흡하다.
MVC 패턴을 사용한다고 해봤는데, 확실히 내가 객체간의 책임에 대해서 제대로 인지하지 못하고 있으니, 메서드 하나를 작성하는 데에도 굉장히 골치가 아프고 힘이 든다.

(cherry picked from commit 868e102a4078b23819fede9a93fbc4fb2acf30ef)

sungik-choi committed:
Docs: 객체 별 모듈 분리

모델, 뷰, 컨트롤러를 개별 파일로 분리하고 main에 import 시키도록 코드 변경.

(cherry picked from commit 99ae7c44f08499281733c84f00581e124e389056)

sungik-choi committed:
Style: 아이콘 파란색 테두리 수정

(cherry picked from commit 003de9bb5fd27ccec35b7b8a6015eac4abb3697a)

DongHyunKims commented on resources/js/controller.js:
```diff
@@ -0,0 +1,75 @@
+
+SearchController.prototype.keyDownCallback = function keyDownCallback(evt, index) {
+  let listIndex = index;
+  const [ARROW_UP, ARROW_DOWN, ENTER] = ["ArrowUp", "ArrowDown", "Enter"];
```
Destructuring을 잘 사용하셨네요~ 👍

DongHyunKims commented on resources/js/controller.js:
```diff
@@ -0,0 +1,75 @@
+  this.searchView.templateList(wordsFromModel);
+  if (wordsFromModel.length > 0) this.searchView.addClass();
+  else this.searchView.removeClass();
+};
```
이 함수 내부에서 사용되는 것이 모두 view에 관련된 것인데 controller에서 사용하는 이유가 있을까요? 
그리고 패턴을 적용해보는 연습을 하는것도 좋지만 과도하게 적용하게 되면 가독성이 어려운 코드가 될수 있어요! 항상 주의해서 사용하면 좋을것 같아요~

sungik-choi replied:
명령을 내리는 건 다 controller에서 해야한다고 책임을 정해두고, 매개변수를 model로 부터 받아와서 view에게 일을 시키는 형태인데... 말씀하신대로 이 부분에 대해 고민이 많았습니다. controller가 매개변수로 받아온 데이터를 view가 해석할 수 있는 형식으로 변환해주는 것도 아니고, 그대로 model의 데이터를 view에 넣는 형식이라 말씀하신 것처럼 가독성만 떨어지는 것 같습니다.

DongHyunKims commented on resources/js/model.js:
```diff
@@ -0,0 +1,32 @@
+  this.src = src;
+};
+
+SearchModel.prototype.getList = function getList(str, fn) {
```
보통 `get` 간단히 데이터를 가져올때 사용해요~ 예를 들어 배열에서 하나의 데이터를 가져오는 정도의 역할에서요~ 보통 서버와 통신한다면 Fetch를 사용하는게 일반적이니 fetchList로 작성해도 무난 할것 같아요~

그리고 fn 보다 callback과 같은 이름으로 parmater로 작성하면 더욱 가독성이 좋아 지겠네요~

DongHyunKims commented on resources/js/model.js:
```diff
@@ -0,0 +1,32 @@
+
+SearchModel.prototype.getList = function getList(str, fn) {
+  fetch(this.src)
+    .then(response => {
+      return response.json();
+    })
```
이렇게 반복이 예상되는 코드는 fetch를 wrapping하여 간편히 만들어 보는 연습을 하면 좋아요~

DongHyunKims commented on resources/js/model.js:
```diff
@@ -0,0 +1,32 @@
+    .then(data => {
+      const regex = new RegExp(`^(${str})(?:W*)(.*)?$`);
+      const matchedWord = this.matchWord(regex, data.list);
+      fn(matchedWord);
```
fn이 function 인지 확인 하거나 존재 하는지 확인하는 방어코드가 있으면 좋겠네요~

DongHyunKims commented on resources/js/model.js:
```diff
@@ -0,0 +1,32 @@
+        restSpell: eachWord.match(regex)[1] || "",
+      });
+    }
+  });
```
정규표현식 사용해 보는 시도는 좋아보여요! 처음 같은 단어 매칭하고 해당 단어를 하이라이팅 해주는 부분인것 같은데 매칭되지 않는 부분을 bold 처리하는 이유가 있을까요? 그리고 `eachWord.match(regex)`을 변수로 할당해주고 중복을 줄여보면 좋겠네요~

sungik-choi replied:
기획서에는 해당 단어를 하이라이팅 해주라고 작성되있었는데, 기획서의 아마존 예시 사진과 구글 같은 검색엔진에서는 대부분 해당 단어를 제외한 나머지 부분을 하이라이팅 해주는 것 같아 그 방식으로 처리해봤습니다.

DongHyunKims reviewed (commented):
한 주동안  정말 고생 많이 하셨어요! 패턴도 고민 해보면서 열심히 하는 모습 보기 좋습니다~ 정리한 내용을 다른 분들과도 공유 해보면 좋을것 같네요~

DongHyunKims commented on resources/js/model.js:
```diff
@@ -0,0 +1,32 @@
+      });
+    }
+  });
+  if (word.length >= MAX_SIZE) word.length = MAX_SIZE;
```
보통 array의 기본 property인 length의 값은 변경하지 않지 않아요! 이렇게 작성한 이유가 있을까요?

sungik-choi replied:
<https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Array/length>

이 부분을 참고해서, 자동검색완성에 출력되는 최대 단어 갯수를 10개로 제한하고 싶어서 배열 단축의 용도로 사용했습니다. 어떤 방식으로 작성하는게 맞는 방식일까요..?

crongro replied:
배열의 함수를 쓰는게 좀더 좋겠네요.

sungik-choi commented:
리뷰 감사합니다!

sungik-choi committed:
Feat/Fix: 코드리뷰 내용 수정 및 인덱스 관련 오류 수정

- view.js

인덱스 변수를 controller 내부 메서드의 지역변수가 아닌, view 내부 생성자의 프로퍼티로 둬서 관리하기로 했다.  clearList() 라는 메서드는 active 클래스를 삭제하고, 동시에 내부 리스트 초기화, 인덱스 초기화를 담당하게 된다. 리스트를 초기화해서, 자동완성 리스트가 보이지 않는다면 innerHTML 리스트 요소로 접근이 불가능하게, 그리고 다음 자동완성 리스트를 패치해왔을때 다시 인덱스를 초기화하게 변경했다. 잘 작동은 하는데, 아쉬운점은 기존에 addClass/removeClass가 둘 다 사용되었는데 이 메서드가 추가된 후로는 addClass/clearList 로 사용돼서 일관성이 떨어지는 느낌이다.

- controller.js

매개변수로 index를 받는 여러 메서드들에서 index를 삭제하고, view의 listIndex를 변경하는 방식으로 변경했다. 그래서 index가 사용되던 부분들은 코드가 다소 길어졌다. 기존에 removeClass() 사용 부분에 clearList()를 사용하면서 매번 listIndex를 -1로 초기화시켜주도록 했다.

- model.js

리뷰 받은 내용을 바탕으로 메서드 명을 변경하고, 방어코드를 작성했다. 중복 코드는 계속해서 염두해두면서 중복을 줄여나가도록 계속 신경써야겠다.

(cherry picked from commit 739a464d01d68cac7105412a7b55580b92a4bf44)

crongro commented:
> * MVC 패턴에 더해서 `왜 이런식의 코드를 작성했는지?`
이렇게 의문점을 갖는게 좋죠. 고민은 계속하고, 비판적으로 받아들이셔야 하고요.(과연 필요한가? 이게 ? 이런식)
모든패턴은 패턴 그 자체로 의미가 별로 없어요. 
코드에 변화가 생기고, view를 재사용하는 등의 복잡한 상황이 닥쳐야 의미가 생겨요.

crongro commented:
여기까지 머지합니다!
